### DISCUSSION TITLE
MEGA-BAS input Transition Counts

### DISCUSSION THREAD

**Discussion by orrious** [tagyoureit/nodejs-poolController#536] [REM]
How do you configure a MEGA-BAS input to act as a counter?

Should be able to set read frequency, reset frequency, and then feed that into a generic device that tabulates flow rate per period into total consumption.

**Comment by orrious** [comment]
For context, I have a flow meter on the auto-top off for the pool.  I can calculate usage based on counts per second over time.  You can't just let the counter go and check it every 30 seconds, you need to check it every 500ms.

**Reply by rstrouse** [reply]
I wrote some code to read the edge count on the BAS but never popped it into the repo.  The BAS actually keeps this counter internally on its own interrupt.  Keep in mind it is only for the dry contact settings.  I will add the UI and emits next week for it.

**Reply by orrious** [reply]
You can `countrd` and `countrst`.  Most folks are going to purchase a "flow meter" not a "water meter", so the counter is going to mean different things.  A water meter is going to associate a quantity of water with each "tick".  If that's the case then it doesn't matter when you read the counter as it's a fixed quantity.

Flow meters on the other hand require a different approach.  To get volume you need flow over time and then add that to the total volume.  

I would have the BAS `Feed` the count to a Generic Device `Flow Totalizer`every second.  Each flow meter has a different tick to time flow rate, so end users should be able to enter that value.  The totalizer would get the total count and store it, with the time it got it.  Then as the next count came in, it would subtract the old count from the new count to get the current time count.  Perform count / time calculation and store the total volume for that time and then add it to total volume.  

Eventually the counter is going to either go back to zero or you will need to `countrst`

**Reply by rstrouse** [reply]
External pulse counts are not appropriate for totalizer meters.  The flow meters we may encounter are not really all that accurate and rollover on the pulse counter is very easy to detect and manage.  Even when you average them over longer periods of time you can still fall within the margins of what one of these turbine based pulse read meters are capable of.

The information emitted on the output will include a delta count and time in ms between the reads.

**Reply by orrious** [reply]
Most flow meters I've seen rely on frequency, which is pulses / sec.  

This is the formula for the flow meter I have:   F=5.5 * Q (L / Min)
https://www.digiten.shop/products/digiten-g3-4-water-flow-hall-sensor-switch-flow-meter-1-60l-min

I was skeptical that the two would average out, but at least with this example it works either way.

Take for example a time period of 1 minute.
0-9 seconds are at 330hz or 330 pulses a second.  330/5.5=60L minute.  60Lm/60secs = 1L second for 10 seconds = 10L
10-19 seconds are at 200hz or 200 pulses a second.  200/5.5=36.36Lm / 60 seconds = 0.60Ls for 10 seconds = 6L
20-49 seconds have no pulses / no flow
50-59 seconds are at 50hz or 50 pulses a second.  50/5.5=9.09Lm / 60 seconds =  0.15Ls for 10 seconds = 1.5L

There are 5800 total pulses over 60 seconds = 96.66hz / 5.5 = 17.57Lm

I guess you would need only supply the totalizer with the flow rate formula for your flow meter and the counts.

Ironically after hooking the flow meter to the MEGA-BAS input, It will not count it.

**Comment by orrious** [comment]
Will the MEGA-BAS support pulse counting from a digital flow meter or does it only do dry contact counting?

**Reply by rstrouse** [reply]
It appears to only do dry contact counting.   Often things like this get tied to a 4-20mA controller totalizer but a poor mans version may simply be triggering on a GPIO interrupt.  A bit of mods in REM should suffice given it measures rising and lowering edge on the interrupt to get square wave.

**Reply by orrious** [reply]
Depends on the frequency. RPi is not well suited to do any type of reliable frequency counting. That was the whole reason I went with an Ardunio Mega as my board in the first place.  I though that since the MEGA-BAS had its own MCU it would be well suited for pulse counting. 

I’ve been trying to avoid this at all cost, but I may just integrate my Mega board into REM. It’s already I2C and will give me a good project to learn njs on.

**Reply by rstrouse** [reply]
Good idea I have about 10 of them floating around.

**Reply by orrious** [reply]
My board is already modular.  I can generate a config per use case or more preferably have a fixed configuration.  With 16 analog inputs and 54 GPIOs that can be used for input, output, pulse counters, etc.  It would be much easier to have them all pre-defined.

My current pinmode types are:
thermistor - 10k
voltage or adc value analog
digital in
digital out
pwm out
intelibrite
pulse counter

Can you give me a swag on the number of each you would do?

If a fixed configuration isn't preferred, I could make the board programmable over I2C and save to flash to persist reboots.  That may be the next iteration.

**Reply by rstrouse** [reply]
Well I missed your questions.  Here is what I would expect for I/O that would cover single, shared, and dual body configurations.

Analog:
4 for 10k Thermistor resistance - (waterSensor1, waterSensor2, solar, air).  This allows for dual body configs.
2 for pressure sensors ADC value
1 for board voltage

Digital Input
2 for flow sensors

Pulse Count
1 for fill line.  This should probably be a totalizer arrangement for such a thing with rollover detection.  Simply a continuous count up from start.
2 for flow rate.  The idea for such a sensor would be to get instantaneous flow rate.   Simply pulse count/interval.

Digital Output
Digital outputs are an interesting duck depending on how they are implemented.  In all cases they should be latched to some power off state and actually treated differently depending on whether the digital out is for valves or circuits.   For circuits, one might simply drive a transistor to switch 24vdc on contactors drawing a small amount of power.  I am not sure whether that could drive a servo motor for things like chemistry pumps though.

The other thing that is a bit of a wrinkle are valve actuators.  These are best to drive with a relay since there is an on-off-on state for them and one of those on states are the default state.   Power to at least one of the valve legs at all times are required.

PWM
For the most part I don't see any benefit for PWM on that level.  Any fixture that would need something like that would need additional hardware like a triac at the power levels of the device.  Maybe to drive indicator LEDs on the board itself?

Just my ramblings
